lib/Models/RegionProvider.js

```diff
@@ -0,0 +1,405 @@
+        } 
+    }
+};
+
+/** 
+ * Given an entry from the region mapping config, contact our WFS server to get the list of all IDs of that type 
+ * @returns {Object} Promise that returns true if IDs were loaded for the first time, or false if already loaded.
+*/
+
+RegionProvider.prototype.loadRegionIDs = function() {
+    if (this.regions.length > 0) {
+        return when(false); // already loaded, so return insta-promise.
+    }
+    if (this.server === undefined) {
+        throw (new DeveloperError('No server for region mapping defined: ' + this.regionType));
+    }            
+    var that = this;
+
+    var baseuri = URI(this.server).addQuery({ 
+        service: 'wfs',
+        version: '2.0',
+        request: 'getPropertyValue',
+        typenames: this.layerName});
+
+    // get the list of IDs that we will attempt to match against for this column
+    var p = [];
+    var url = baseuri.addSearch('valueReference', this.regionProp).toString();
+    p.push(loadText(corsProxy.getURL(url))
+        .then(function(xml) {
+            that.loadRegionsFromXML(xml);
+        }).otherwise(function(err) {
+            console.log(err);
+            throw(err);
+        }));
+    // if this column might be ambiguous then fetch the disambiguating values for each column as well (usually State)
+    var dab = this.disambigColumns;
+    if (defined(dab)) {
+        
+        url = baseuri.addQuery('valueReference' + dab[0].regionProp).toString();
+        p.push(loadText(corsProxy.getURL(url)).then(function(xml) {
+            that.loadRegionsFromXML(xml, dab[0].regionProp);
+        }));
+    }
+    return when.all(p).yield(true);
+};
+
+
+function applyReplacements(s, replacements) {
+// replacements: [ [ regex, replacement], ... ] 
+// replacement can contain '$1' etc.
+    var r = s.trim().toLowerCase(); // just in case...
+
+    if (replacements === undefined || replacements.length === 0) {
+        return r;
+    }
+
+    var cachekey = s + '%' + replacements.reduce(function(key, item) { 
+        return key + item[0] + '%' + (item[1] ? item[1] : '') + '%'; 
+        }, '');
+    if (applyReplacements.cache[cachekey] !== undefined) {
+        return applyReplacements.cache[cachekey];
+    }
+    
+    replacements.forEach(function(rep) {
+        r = r.replace(new RegExp(rep[0].toLowerCase(), 'gi'), rep[1].toLowerCase());
+    });
+    applyReplacements.cache[s] =r;
+    return r;
+}
+applyReplacements.cache = {};
+
+
+/* Checks if a given region code matches a given ID after applying regexs and additional disambiguation lookups */
+RegionProvider.prototype.codeMatchesRegionID = function(id, code, disambigId, disambigCode) { //, row, idnum, csvItem ) {
+    code = String(code).toLowerCase(); //## should be happening upstream somewhere
+    id = String(id).toLowerCase(); 
+    if (applyReplacements(id, this.serverReplacements) !== applyReplacements(code, this.dataReplacements)) {
```

Comment:
>Couldn't you do the serverReplacements once instead of doing it while testing each code?

Ah, yep. Logical next refinement. Basically just moving the cache up a level, so it's a simple lookup. Nice catch.